### ISSUE TITLE
DataLoaderがバグっている？

### ISSUE THREAD

**Issue by yaneurao** [TadaoYamaoka/DeepLearningShogi#39]
確信はないのですが、DataLoaderがバグっている可能性ってないでしょうか？

DataLoaderのcommitをmergeして、同じ教師で同じパラメータで学習をしても、従来のdlshogiのtrain_rl_policy_with_value_using_hcpe_bootstrap.pyのログとlossの値が全く違うものが出来てしまいます。

パラメータ
2021/03/04 22:22:11    INFO    network=resnet10_swish
2021/03/04 22:22:11    INFO    batchsize=8192
2021/03/04 22:22:11    INFO    MomentumSGD(lr=1e-05)
2021/03/04 22:22:11    INFO    WeightDecay(rate=0.0001)
2021/03/04 22:22:11    INFO    entropy regularization coeff=0.001
2021/03/04 22:22:11    INFO    val_lambda=0.333
2021/03/04 22:22:12    INFO    use amp

従来の学習
2021/03/04 09:34:56    INFO    epoch = 19, iteration = 2971000, loss = 0.58594222, 0.39444258, 0.59953060, 1.04867912, test loss = 0.85237104, 0.49494448, 0.59407979, 1.38032746, test accuracy = 0.50000000, 0.74687500

読み込みの高速化まで適用した場合の学習
2021/03/04 22:33:50    INFO    epoch = 19, iteration = 2971000, loss = 2.12280966, 0.41511919, 0.60691084, 2.60179547, test loss = 2.00746894, 0.51670283, 0.59571248, 2.55048203, test accuracy = 0.50000000, 0.74218750

// GPUでデータを消費する前にpre_fetchで次のバッチが来てデータを上書きしている？

見た感じaccuracyは同じぐらいなので学習には影響はないのでしょうか？
// DataLoaderクラスのpre_fetchメソッドが実質的にはmini_batchメソッドを呼び出していて、mini_batchメソッドは現在のバッチの利用状況に関係なく次のバッチをGPUに転送してしまうので、学習データが壊れているケースが存在する？

**Comment by yaneurao** [comment]
lossの出力がおかしいと上で書いたのは、ビルドに失敗していた模様。(たややんさんが)
// 学習がうまくできているのかはいま検証中です。お騒がせして申し訳有りません。

たややん今日 08:00
大変申し訳ないです、確認したら、私がcppshogiの再ビルドをミスっていただけの可能性が浮上しました…新しい学習部を使ってもログは正常に出力されるようになったかもしれません。

**Comment by TadaoYamaoka** [comment]
git pullして、ビルドしただけだと、Visual Studioがcppshogiを自動で差分のビルドをしてくれないようです。
cppshogiを明示的にリビルドしていないと上記のような状態になります。

**Comment by TadaoYamaoka** [comment]
なお、pre_fetchでは、CPU側のメモリを編集していて、前のバッチはGPUに転送済みなので、上書きは発生しないはずです。

**Comment by windfall-shogi** [comment]
前のバッチはGPUに転送済みですが、そのデータで学習を行う前に次のバッチのpre_fetchが完了してしまう可能性が低確率ながらあると思います。
pre_fetchの処理の中身はmini_batchで、これはGPUへの転送まで行うので未使用の前のバッチを上書きする可能性はあると思います。

**Comment by TadaoYamaoka** [comment]
https://github.com/TadaoYamaoka/DeepLearningShogi/blob/274bd9447e964706bea96f807cedebd19e03e7a7/dlshogi/train_rl_policy_with_value_using_hcpe_bootstrap.py#L167
ここで、pre_fetchしたミニバッチのGPUへの転送が確実に終わっているので、次のpre_fetchを行っても上書きされることはあり得ないです。

**Comment by yaneurao** [comment]
学習、ちゃんとできているようでした。お騒がせしました。🙇
